### DISCUSSION TITLE
"focus" on shapes in ui

### DISCUSSION THREAD

**Discussion by Jonne-G** [emilk/egui#3035] [Q&A]
Hi, 
I am writing a small application for tinkering with gears and am currently working on some quality of life features.
One of these is centering the "viewport" on the currently placed gears.

Essentially I would want the user to press a button and go from this:
![gear_toy_2uiHs7igUY](https://github.com/emilk/egui/assets/60442130/df139821-869b-4aa4-b34f-e8c82892fa8f)

to this:
![gear_toy_iVv7sbaPpI](https://github.com/emilk/egui/assets/60442130/90d78ea2-3e18-4473-85f3-99f15d81add3)

For context, I am working off the "fractal clock" example (https://www.egui.rs/#Clock) for drawing the shapes,
with a simple position offset to make panning around work.

Is there a recommended method for doing something like this?

**Comment by JumpyLionnn** [comment]
If you are talking about the scaling you can just make the rect of the shapes bigger and do some calculations to fit the rect.
But if you are asking how to manage state here, there are a couple of ideas that come to my mind.
1. You can simply store the state in a variable. This is not the most "immediate" style but it gets the job done and its easy to do.
2. if you allocated space for the shapes you can use the response you get from the allocation and check if it is focused. using the response you can request focus and you can also store the ui's id to focus it using the `ctx.memory()`, see https://docs.rs/egui/latest/egui/struct.Memory.html#method.request_focus
3. you can also manage the state yourself in "immediate" style using `ctx.data()`. see https://docs.rs/egui/latest/egui/struct.Context.html#method.data.

I hope it helped you :)
